cmd/dagger/functions.go

```diff
@@ -54,6 +54,13 @@ available functions.
 		"`",
 	),
 	Execute: func(fc *FuncCommand, cmd *cobra.Command) error {
+		// Manually check for --help in the arguments
+		for _, arg := range cmd.Flags().Args() {
+			if arg == "--help" || arg == "-h" {
+				return cmd.Help()
+			}
+		}
```

Comment:
If you look at what `cmd.Flags().Args()` does, it should return the command's arguments without flags. The reason it's appearing here is that `FuncCommand` disables interspersed flag parsing. The better solution is to handle this in `FuncCommand` so that flags are parsed before getting to this point. 

## Technical explanation

<details><summary>Why a <code>FuncCommand</code> abstration?</summary>

Here’s a bit of context. `FuncCommand` was created to support several commands that work like `dagger call`:
- `dagger download`
- `dagger up`
- `dagger shell`
- `dagger print` (eventually)

`dagger functions` was special in that it requires most of the same boilerplate, but without the dynamic generation of Cobra commands based on function chaining. So today we’re in a weird place where we have a `FuncCommand` abstraction for just 1 case and 50% of another case (without the subcommands).

We could simplify to just have a common bootstrapping function but the time to refactor this isn’t worth it when there’s other priorities. Besides, it actually keeps all the chaining logic away from the specific `call` implementation so it’s easier to see what `call` does specifically. Same for `functions`. 
</details>

In `dagger call`, we need to disable interspersed flag parsing because it would try to parse a flag to a command that isn’t created yet. These commands are created dynamically and recursively as long as the next positional argument matches an available function from the object that the current function returns.

During this process, each subcommand will have its own set of flags, which is going to be parsed during this creation process. That’s what *traverse children* does.

For `functions` though, it doesn’t go forward with creating subcommands, so by the time `Execute`  runs, any flags after a positional argument haven’t been parsed. But that [can be fixed](https://github.com/dagger/dagger/pull/7187).

Reply:
Thanks for the context. I was confused at first with the FuncCommand being halfway executed when `Execute` is present, but the context indeed makes sense 🙏 

> This is best handled by native flag parsing which knows which flags are defined, rather than hardcode these values. But it should be done in FuncCommand for better separation of concerns.
Yes, agreed. I tried but failed. Your implementation is insightful, I would have never found out 🙏